src/window.rs

```diff
@@ -0,0 +1,138 @@
+use core::marker::PhantomData;
+use core::f64;
+
+use {Sample, FloatSample};
+use signal;
+use signal::{Phase, ConstHz, Step};
+use conv::{FromSample, ToSample};
+use frame::Frame;
+
+// Using types instead of enum allows for static dispatch
+pub trait WindowType<S: Sample, T: Step> {
+    fn at_phase(phase: S) -> S;
+}
+
+pub struct HanningWindow;
+
+impl<S: Sample + ToSample<f64> + FromSample<f64>, T: Step> WindowType<S, T> for HanningWindow {
+    fn at_phase(phase: S) -> S {
+        let v = phase.to_sample::<f64>() * f64::consts::PI * 2.;
+        (0.5f64 * (1f64 - v.cos())).to_sample::<S>()
+    }
+}
+
+pub struct RectangleWindow;
+
+impl<S: Sample + FromSample<f64>, T: Step> WindowType<S, T> for RectangleWindow {
+    #[allow(unused)]
+    fn at_phase(phase: S) -> S {
+        (1.).to_sample::<S>()
+    }
+}
+
+pub struct Window<S: Sample, T: Step, F: Frame<Sample=S>, WT: WindowType<S, T>> {
+    pub phase: Phase<ConstHz>,
+    stype: PhantomData<S>,
+    ttype: PhantomData<T>,
+    ftype: PhantomData<F>,
+    wttype: PhantomData<WT>
+}
+
+impl<S: Sample, F: Frame<Sample=S>, WT: WindowType<S, ConstHz>> Window<S, ConstHz, F, WT> {
```

Comment:
If `Window` is generic over it’s phase and step, perhaps it should be passed to the new parameter, rather than having `Window::new` be only implemented for `Window` where T is ConstHz? Alternatively, can we just assume that the user always wants ConstHz, and remove the type parameter altogether in favour of `Window` having a `ConstHz` field?